edge-util/src/Microsoft.Azure.Devices.Edge.Storage/InMemoryDbStoreProvider.cs

```diff
@@ -1,18 +1,100 @@
 // Copyright (c) Microsoft. All rights reserved.
 namespace Microsoft.Azure.Devices.Edge.Storage
 {
+    using System;
     using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
     using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Converters;
 
     public class InMemoryDbStoreProvider : IDbStoreProvider
     {
+        const string BackupMetadataFileName = "meta.json";
         const string DefaultPartitionName = "$Default";
         readonly ConcurrentDictionary<string, IDbStore> partitionDbStoreDictionary = new ConcurrentDictionary<string, IDbStore>();
+        readonly Option<string> backupPath;
+        readonly bool useBackupAndRestore;
+
+        public InMemoryDbStoreProvider()
+            : this(Option.None<string>(), false)
+        {
+        }
+
+        public InMemoryDbStoreProvider(Option<string> backupPath, bool useBackupAndRestore)
+        {
+            this.backupPath = backupPath;
+            this.useBackupAndRestore = useBackupAndRestore;
+
+            // Restore from a previous backup if enabled.
+            if (useBackupAndRestore)
+            {
+                string backupPathValue = this.backupPath.Expect(() => new ArgumentException($"The value of {nameof(backupPath)} needs to be specified if backup and restore is enabled."));
+                Preconditions.CheckNonWhiteSpace(backupPathValue, nameof(backupPath));
+
+                this.RestoreDb(backupPathValue);
+            }
+        }
+
+        private void RestoreDb(string backupPath)
+        {
+            string backupMetadataFilePath = Path.Combine(backupPath, BackupMetadataFileName);
+
+            if (!File.Exists(backupMetadataFilePath))
+            {
+                Events.NoBackupsForRestore();
+                return;
+            }
+
+            try
+            {
+                using (StreamReader file = File.OpenText(backupMetadataFilePath))
+                {
+                    JsonSerializer serializer = new JsonSerializer();
+                    BackupMetadataList backupMetadataList = (BackupMetadataList)serializer.Deserialize(file, typeof(BackupMetadataList));
```

Comment:
Why do you have a list of metadata here? Shouldn't it be only one? (per your design doc)

Reply:
I have a list here so as to support multiple backups in the future with ease. It wasn't adding a whole lot of bloat to the design so I decided on doing that. The design doc does indicate that we are keeping a list